File: update-package-via-manifest/src/providers/github-tags.ts
import * as core from "@actions/core"
import * as github from "@actions/github"
import { GitHub } from "@actions/github/lib/utils"

import { IUpdateProvider } from "../index"

interface IManifestData {
	repo: string
}

export class GitHubTagUpdateProvider implements IUpdateProvider {
	private lastTagHTMLLink: string | undefined
	private octokit: InstanceType<typeof GitHub>

	constructor() {
		this.octokit = github.getOctokit(core.getInput("github-token"))
	}

	async latestVersion(manifestData: IManifestData): Promise<string | undefined> {
		const split = manifestData.repo.split("/")
		const owner = split[0]
		const repoName = split[1]

		if (owner === undefined || repoName === undefined) {
			return undefined
		}

		let latestTag: string | undefined;
		try {
			latestTag = await this.getLatestTag(owner, repoName)
		} catch (e: any) {
			core.warning(e)
			return undefined
		}

		if (latestTag === undefined) {
			return undefined
		}

		this.lastTagHTMLLink = `https://github.com/${owner}/${repoName}/releases/tag/${latestTag}`

		return latestTag.replace(/^v/m, "") // Make sure any leading v's are removed
	}

	async updateData(_: IManifestData): Promise<{
		updateChecksums: boolean
		prContent?: string | undefined
		sourceArray?: string[] | undefined
		source_x86_64?: string[] | undefined
		source_i686?: string[] | undefined
		source_aarch64?: string[] | undefined
		source_armv7h?: string[] | undefined
	} | undefined> {
		if (this.lastTagHTMLLink === undefined) {
			return undefined
		}

		return {
			updateChecksums: true,
			prContent: `__GitHub Tag Link:__ [${this.lastTagHTMLLink}](${this.lastTagHTMLLink})`
		}
	}

	async getLatestTag(owner: string, repoName: string): Promise<string | undefined> {
		const resp = await this.octokit.rest.repos.listTags({
			owner: owner,
			repo: repoName,
		})

		if (resp.status !== 200) {
			throw new Error(`invalid HTTP response code: ${resp.status}`)
		}

		if (resp.data.length === 0) {
			throw new Error("received a 0-length array when listing Git tags")
		}

		resp.data.sort()

		if (resp.data[0] === undefined) {
			throw new Error("Something's broken. Don't know what, but it's broken.")
		}

		return resp.data[0].name
	}
}

File: update-package-via-manifest/src/providers/github_tags.rs
use std::{any::Any, env};

use anyhow::anyhow;
use serde::Deserialize;

use crate::providers::{UpdateData, UpdateProvider};

pub struct GitHubTagsProvider {
    last_tag_html_link: Option<String>,
}

impl UpdateProvider for GitHubTagsProvider {
    fn latest_version(&mut self, manifest_data: &Box<dyn Any>) -> anyhow::Result<String> {
        let data = manifest_data
            .downcast_ref::<GHTagsData>()
            .ok_or(anyhow!("Failed to downcast manifest_data to a GHTagsData"))?;

        let split: Vec<&str> = data.repo.split('/').collect();
        let owner = split
            .first()
            .ok_or(anyhow!("Couldn't extract owner from manifest"))?;
        let repo_name = split
            .get(1)
            .ok_or(anyhow!("Couldn't extract repo from manifest"))?;

        let mut tags: Vec<TagResp> = ureq::get(format!(
            "https://api.github.com/repos/{owner}/{repo_name}/tags"
        ))
        .header("Accept", "application/vnd.github+json")
        .header(
            "Authorization",
            format!("Bearer {}", env::var("INPUT_GITHUB-TOKEN")?),
        )
        .header("X-GitHub-Api-Version", "2026-03-10")
        .call()?
        .body_mut()
        .read_json()?;

        tags.sort(); // There's no way sort actually works. This will need to be revisited.

        let mut latest_tag = tags
            .first()
            .ok_or(anyhow!("latest tag could not be found"))?
            .name
            .clone();

        self.last_tag_html_link = Some(format!(
            "https://github.com/{owner}/{repo_name}/releases/tag/{latest_tag}"
        ));

        // Remove a potential leading v (i.e. v1.0.0 -> 1.0.0)
        if latest_tag.starts_with('v') {
            latest_tag.remove(0);
        }

        Ok(latest_tag)
    }

    fn get_update_data(&mut self, _: &Box<dyn Any>) -> anyhow::Result<UpdateData> {
        Ok(UpdateData {
            update_checksums: true,
            pr_content: self
                .last_tag_html_link
                .as_ref()
                .map(|link| format!("__GitHub Tag Link:__ [{link}]({link})")),
            ..Default::default()
        })
    }
}

impl GitHubTagsProvider {
    pub fn new() -> Self {
        Self {
            last_tag_html_link: None,
        }
    }
}

#[derive(Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct GHTagsData {
    pub repo: String,
}

#[derive(Deserialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
struct TagResp {
    name: String,
}
